app/src/main/java/io/pslab/activity/OscilloscopeActivity.java

```diff
@@ -350,154 +352,35 @@ public void run() {
-                            synchronized (lock) {
-                                try {
-                                    lock.wait();
-                                } catch (InterruptedException e) {
-                                    e.printStackTrace();
+                        } else {
+                            if (scienceLab.isConnected()) {
+                                if (isCH1Selected) {
+                                    channels.add(CHANNEL.CH1.toString());
                                 }
-                            }
-                        }
-
-                        if (scienceLab.isConnected() && isCH2Selected && isAudioInputSelected && !isCH3Selected && !isCH1Selected && !isXYPlotSelected) {
-                            captureTask = new CaptureTask();
-                            captureTask.execute(CHANNEL.CH2.toString(), CHANNEL.MIC.toString());
-                            synchronized (lock) {
-                                try {
-                                    lock.wait();
-                                } catch (InterruptedException e) {
-                                    e.printStackTrace();
+                                if (isCH2Selected) {
+                                    channels.add(CHANNEL.CH2.toString());
                                 }
-                            }
-                        }
-
-                        if (scienceLab.isConnected() && isCH3Selected && isAudioInputSelected && !isCH2Selected && !isCH1Selected && !isXYPlotSelected) {
-                            captureTask = new CaptureTask();
-                            captureTask.execute(CHANNEL.CH3.toString(), CHANNEL.MIC.toString());
-                            synchronized (lock) {
-                                try {
-                                    lock.wait();
-                                } catch (InterruptedException e) {
-                                    e.printStackTrace();
+                                if (isCH3Selected) {
+                                    channels.add(CHANNEL.CH3.toString());
                                 }
                             }
-                        }
-
-                        if (scienceLab.isConnected() && isCH1Selected && isCH2Selected && isCH3Selected && !isAudioInputSelected && !isXYPlotSelected) {
-                            captureTask = new CaptureTask();
-                            captureTask.execute(CHANNEL.CH1.toString(), CHANNEL.CH2.toString(), CHANNEL.CH3.toString());
-                            synchronized (lock) {
-                                try {
-                                    lock.wait();
-                                } catch (InterruptedException e) {
-                                    e.printStackTrace();
-                                }
+                            if (isAudioInputSelected && isInBuiltMicSelected || (scienceLab.isConnected() && isMICSelected)) {
```

Comment:
@marcnause We may have to improve this condition/add more conditions for handling the In-Built MIC with this logic. If the _PSLab_ device is connected and the In-Built MIC feature of the phone is toggled _ON_ using the radio button, it won't get toggled _OFF_ at all by unchecking the radio button, and would keep running forever. Could you please once test and confirm if this issue exists ?

Reply:
You are right, the issue exists!

Reply:
@AsCress I think this fix was fixed (😉) by the changes you introduced in https://github.com/fossasia/pslab-android/pull/2529. I  did some tests and I *think* it works as intended now. Can you confirm?

Reply:
@marcnause Yes you are correct :))